DOC: Makeover for Sphinx pages

bjnath opened the pull request:
Larger type, clearer sidebars, and a more modern look, extending the #16988 update.

New:

![image](https://user-images.githubusercontent.com/6691888/89589255-47df0d80-d813-11ea-8e01-b173173c4344.png)

Previous:

![image](https://user-images.githubusercontent.com/6691888/89589264-4dd4ee80-d813-11ea-8d83-e7d5a4e70bb4.png)

Looking forward to comments (@seiko2plus, dig in!). 

This is again a change just to `layout.html`. I tried linking to a numpy.css stylesheet in a static directory, but I found that when the pages were built my relative link to the stylesheet did not always end up pointing to the same path, for reasons I can't explain.

I'll post a link to the CircleCI build object once the build completes.

bjnath committed:
DOC: New CSS style for NumPy Sphinx pages

bjnath committed:
DOC: Updated look for Sphinx pages

Extending the makeover of the Sphinx pages for larger type, clearer
sidebars, and a more modern look.

bjnath commented:
Finished product is in https://15031-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

rgommers commented:
Ooh very nice:) Thanks @bjnath, this makes a big difference - looks much more 2020 rather than 2010.

If you're up for another tweak, the yellow in the signature boxes could use an update (not sure what, maybe some blue-ish color in the website color palette?):

<img width="598" alt="image" src="https://user-images.githubusercontent.com/98330/89656882-8553a180-d8cc-11ea-8a19-c113e5d259fb.png">

eric-wieser commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,137 @@
+    text-align: right;
+    letter-spacing: -2px;
+    word-spacing: .03em;
+    margin-top: 9%;
+    margin-bottom: 10%;
```
Using percentages for margins results in very weird behavior when the window is resized - percentages are always a fraction of the width, not the height - so this makes the page stretch down when it is widened!

eric-wieser replied:
I'd recommend using `em` units, since those scale with the text size not the window width.

eric-wieser commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,137 @@
+  @import url('https://fonts.googleapis.com/css2?family=Lato:ital,wght@0,400;0,700;0,900;1,400;1,700;1,900&family=Open+Sans:ital,wght@0,400;0,600;1,400;1,600&display=swap');
```
```suggestion
@import url('https://fonts.googleapis.com/css2?family=Lato:ital,wght@0,400;0,700;0,900;1,400;1,700;1,900&family=Open+Sans:ital,wght@0,400;0,600;1,400;1,600&display=swap');

```

eric-wieser commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,137 @@
+  margin-left: 5%;
+}
+
+body {
+    font-family: 'Open Sans', sans-serif;
+    font-size: 16px;
+    line-height: 150%;
+}
```
```suggestion
body {
  font-family: 'Open Sans', sans-serif;
  font-size: 16px;
  line-height: 150%;
}

```

eric-wieser commented on doc/source/_templates/layout.html:
```diff
@@ -1,25 +1,145 @@
 {% extends "!layout.html" %}
 
 {%- block extrahead %}
+<!-- Was unable to link to a _static stylesheet. Depending on the type of
+page, the relative path wouldn't always go to the same directory. -->
```
Something like this?
```suggestion
<link rel="stylesheet" href="{{ pathto('_static/numpy.css', 1) }}" type="text/css" />
```
from https://github.com/readthedocs/sphinx_rtd_theme/blob/master/sphinx_rtd_theme/layout.html#L24

bjnath replied:
Works perfectly -- thank you! Having the css in the template meant having to rebuild every time I made a style change.

bjnath commented:
@rgommers --

> the yellow in the signature boxes

It's a highlight (doesn't appear if you click on the page directly); a blue wouldn't have made clear it's a highlight -- changed to a diluted version of the mustard in the palette:

![image](https://user-images.githubusercontent.com/6691888/89675477-ffc5f680-d8b7-11ea-8b5c-dce2db60ea4e.png)

Not quite so garish.

bjnath commented:
@rgommers -- Or we can just turn highlighting off.

eric-wieser commented:
Highlighting is useful when coming from a search term, or if there are multiple functions on one page. Can we just disable it for single-function autosummary pages?

bjnath commented:
> disable it for single-function autosummary pages?
That would be logical, but it'd take deeper thinking than I'm up for. I agree, highlighting is useful, and perhaps the color in https://github.com/numpy/numpy/pull/17024#issuecomment-670644989 is a decent compromise.

bjnath committed:
DOC: Add suggested changes to PR #17024

layout.html now reads a static numpy.css file, thanks to @eric-wieser.

At @eric-wieser's suggestion, removed percentage spacing dimensions, added
formatting cleanup.

Per @rgommers, made highlighting color less garish.

eric-wieser commented:
My comment was more a disagreement against total removal. Muting the color strikes me as fine.

bjnath commented:
Incorporated changes -- thank @eric-wieser for your suggestions, particularly the stylesheet save, and @rgommers for your suggestion on highlights.

The bigger titles mean that the names of some routines in the API pages extend off the page; in those cases a horizontal scrollbar appears:
![image](https://user-images.githubusercontent.com/6691888/89680256-9d252880-d8c0-11ea-828d-36606c9ce01b.png)

I'll post the new CircleCI link when it comes around.

bjnath commented:
Latest in https://15035-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

eric-wieser commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,137 @@
+}
+
+div#spc-section-body h1, h2, h3, h4, h5, h6 {
+    font-family: 'Lato', sans-serif;
```
Nit: would be good to pick either 2 or 4 space indents here

bjnath replied:
You're right; fixed.

seiko2plus reviewed (changes requested):
#[1/4] Increase the responsivity and expanding the width.
Reasonable due to increasing font-size and shfiting the main headline to the other side

bjnath committed:
DOC: Formatting changes to PR #17024

Logo section fills page, per @seiko2pro.
Uniform CSS indenting, per @eric-wieser.

bjnath commented:
@eric-wieser: Corrected indents.

@seiko2plus: Increased logo section to fill page; thanks.  Still reluctant to widen the text, even at the increased type size. It's just readable now and I like the white borders.

Wide version:
![image](https://user-images.githubusercontent.com/6691888/89711878-09ed0100-d95b-11ea-8ee6-8dfecc9e1c00.png)

Committed version:

![image](https://user-images.githubusercontent.com/6691888/89711888-196c4a00-d95b-11ea-87ad-c7e34c1befd1.png)


I increased the logo size a little, which I'm already regretting and am changing back.

bjnath committed:
DOC: Revert logo size increase in PR #17024

At 125px it competes with the heading; restored to the earlier 110px.

bjnath commented:
Latest in https://15041-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

eric-wieser commented:
Agreed that the full width is worse - especially on wide monitors

eric-wieser commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,141 @@
+  display: none;
+}
+
+/* Don't show the page title in TOC */
```
Do we want to remove this from the template, rather than just hiding it in the CDs?

bjnath replied:
The two elements I've hidden are the unnecessary "Table of Contents" title and the unnecessary top link that's the page name; what template change would do this instead?

eric-wieser replied:
I've no idea, just thought it might give you more control.

Do you know which template emits the sidebar?

Also, would these changes make more sense in the numpy Sphinx theme, wherever that lives?

bjnath replied:
>do we want to remove this from the template?

I understand you now. I think the partitioning is OK -- all the style changes are in one place, the numpy.css file, and the only template change is to point to that file.

> would these changes make more sense in the numpy Sphinx theme, wherever that lives

My first thought was to do that, but I think we're just a veneer on the scipy theme, which we have as a submodule.

eric-wieser replied:
> but I think we're just a veneer on the scipy theme, which we have as a submodule.

Yes, I was thinking of the abandoned PR below.

If scipy wants similar changes (@rgommers?), then perhaps you should edit the theme directly.

rgommers replied:
I think there are much better Sphinx themes around (e.g. the statsmodels and pandas ones), so longer term I do think we want to get rid of the current theme and just adopt another one. The maintenance cost of a theme is quite high, so sharing the load will be helpful too.

I'm happy with the current approach, adding a `numpy.css`. With the main goal being to improve the most glaring issues (font size, and styling which matches the website better). Other improvements are nice to have, but no need to redesign half the theme I think, because at some point we'll want to get rid of it.

seiko2plus commented:
What I suggested is full width with max-width `1500px` and stretching the reading sections to the max.

The current:
![current](https://user-images.githubusercontent.com/12713707/89716071-54499e80-d9aa-11ea-8c27-c038d1edea87.png)
-----

The suggestion:
![mine](https://user-images.githubusercontent.com/12713707/89716074-57dd2580-d9aa-11ea-9f6f-788fbb7ee3f4.png)

eric-wieser commented:
This likely conflicts with #14648. Are we considering that PR abandoned?

rgommers commented:
> This likely conflicts with #14648. Are we considering that PR abandoned?

Yes I think so.

eric-wieser commented:
So by association, the https://github.com/scipy/scipy-sphinx-theme-v2 repo is defunct?

bjnath committed:
DOC: Float the pages in PR #17024

Incorporates @seiko2plus's changes so
sidebar moves with main text when page is widened.

bjnath commented:
Sorry, @seiko2plus, your styling suggestions are excellent once again and I put them in. You're right, not so good to have the sidebar anchored at the left. Our one difference is max width, which I've put at 900px rather than 1500px.

bjnath commented:
Latest: https://15043-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

seiko2plus commented:
@bjnath, 900px is kinda tight, I feel like 1100px gonna make everybody happy try it locally at least.

bjnath commented:
I've linked **full-size page images** at widths of 900, 1000, and 1100 pixels, so everyone can take a look.

Clicking the link shows a long, skinny image; then click on the image to see the page life-size.

[900px](https://user-images.githubusercontent.com/6691888/89721629-ee5f1600-d9ad-11ea-9f51-2bcabc56879e.png)

[1000px](https://user-images.githubusercontent.com/6691888/89721632-f0c17000-d9ad-11ea-8402-bdaff2697407.png)

[1100px](https://user-images.githubusercontent.com/6691888/89721636-f323ca00-d9ad-11ea-86ff-3f2d3e558550.png)

Here's an [inconclusive link](https://graphicdesign.stackexchange.com/questions/13724/recommended-column-width-for-text-reading-digital-vs-printed) on recommended line lengths.

Character counts of the first line of these pages:
```
900px  78
1000px 88
1100px 94
```

seiko2plus commented:
@bjnath, My concern is to expand figures and code samples on wide screens.
Not sure to be honest about what the line length should be but I daily use Wikipedia
and the line there can normally reach 180-200 chars.
Also on the other side, the user can always turn on the reader view from the browser.

bjnath commented:
@seiko2plus I've noticed that about Wikipedia!

I don't want habits to keep me from seeing a better idea -- let's let this settle for a bit and see if others want to say something. 

Figures and code samples is something I hadn't considered, but I wasn't sure I understood. Code samples should not get too wide because they follow coding-style line-length restrictions. For figures, we probably don't want figures that don't show up well unless the page is expanded. I imagine tables is another example in this genre, but I don't think NumPy docs ordinarily need to display tables with many columns. Did I misunderstand?

Fortunately max width is an easy changeover. It can even be done later if we think we've chosen wrong.

bjnath commented:
Looking at the CI build, I see NEPs were not restyled by this change -- they have just the #16988 changes. 

I'm not clear why that is -- I'd expect they'd either have this change or none.

Their appearance seems OK as-is; they're clearly a separate section, and they look fine with the  #16988 changes.

eric-wieser commented:
I think we should leave line lengths for another PR, rather than piling on increasingly subjective changes in a single PR. My main concern now is whether scipy wants these changes too, and if so if we should just make them directly at scipy/scipy-sphinx-theme

seiko2plus commented on doc/source/_static/numpy.css:
```diff
@@ -0,0 +1,162 @@
+  padding-left: 27px;
+}
+
+div.spc-navbar .nav-pills > li > a {
+  background-color: #4d77cf; /* han blue */
+}
```
```suggestion
div.spc-navbar {
  padding: 14px 0;
  overflow: hidden;
}
div.spc-navbar .nav-pills > li {
  display: flex;
  overflow: hidden;
}
div.spc-navbar .nav-pills > li > a {
  font-size: 14px;
  font-weight: bold;
  padding: 6px 8px;
  -webkit-transition: color .4s;
  transition: color .4s;
}

div.spc-navbar .pull-left {
  width: 70%; /*given enough space to pull-right*/
}
div.spc-navbar .pull-left > li > a,
div.spc-navbar .pull-left > li > a:hover,
div.spc-navbar .pull-left > li > a:focus {
  background: none;
  border: none;
}
div.spc-navbar .pull-left > li > a:hover {
  color: #39f;
}
div.spc-navbar .pull-left > li::after {
  content: '>';
  color: #f1f1f1;
  padding: 6px 0px;
}
div.spc-navbar .pull-left > li:last-child::after{
  content: '';
}
div.spc-navbar .pull-right > li {
  margin: 0 2px;
}
div.spc-navbar .pull-right > li > a {
  background: #fff;
  color: #111;
  font-weight: bold;
  text-transform: capitalize;
}
div.spc-navbar .pull-right > li > a:hover,
div.spc-navbar .pull-right > li > a:focus {
  background: #013243;
  color: #fff;
}

```

seiko2plus reviewed (changes requested):
Alright, we can just avoid talking about line width for now.
#[2/4] These changes aim to modernize the navbar including breadcrumbs and nav-buttons,
the reason behind them is to clarify the sidebar view (table of contents),
so the ideas can be flowing to our minds(or at least my mind) since honestly, I don't like how the sidebar looks right now.

Before
![before](https://user-images.githubusercontent.com/12713707/89742123-6a815880-da97-11ea-846a-041b378e1003.png)

After
![after](https://user-images.githubusercontent.com/12713707/89742096-29894400-da97-11ea-8343-ca599dc9e137.png)

bjnath commented:
>make them directly at scipy/scipy-sphinx-theme

Let's not be doctrinaire.  The PR took me a day to write only because I knew nothing about the scipy theme. If I'd started by trying to understand the theme and making changes bottom-up we'd have pieces all over the floor and still no PR. 

Heed the lessons of #14648. 

> piling on increasingly subjective changes in a single PR

To be clear, nothing's being piled -- it's the value of one number in one line of the stylesheet.  

And increasingly subjective? It's not as if we've been trying on clothes for a week. It's the one esthetic decision we've discussed; @seiko2plus has earned a respectful hearing, and presumably if he feels strongly, and I feel strongly, others also feel strongly; so let's give everyone a moment to speak up. I've provided links to the look over the choices.

bjnath commented:
> piling on increasingly subjective changes

I had not seen the most recent proposals.  Yes, @seiko2plus, maybe the new changes should go in a separate PR.

seiko2plus commented:
@bjnath, I thought I'm here to review the design, not just the code. The layout parts are linked to each other so any real suggestions will lead to massive changes and that looks normal to me.
However, feel free to not accept the changes but in that case, I'm not gonna able to review/suggest changes
on the design only the code.

bjnath commented:
I wrote the PR to have something better than what we had. There was no attempt to establish a new world order or solve everything we hoped to solve.  And I'm not sure the reviewers are expecting that either right now. 

Everything you've suggested to this point has been a slam-dunk obvious improvement. This latest is a new conception and an new kind of discussion.  

Let's agree that what's here now is not everything we hope for, but that it can stand on its own for now while we plot what else needs doing. I expect that discussion will continue for a while.

seiko2plus commented:
> Everything you've suggested to this point has been a slam-dunk obvious improvement. This latest is a new conception and a new kind of discussion.

The lastest changes look obvious to me because I don't like the new sidebar changes and at the same time
I couldn't suggest new changes to the design without improving the navbar since the navbar comes first.

> Let's agree that what's here now is not everything we hope for, but that it can stand on its own while we plot what else needs doing.

Oki, please feel free to not accept the requested changes, and I will move them later to a new pull combined with the other undiscovered suggestions that related to the sidebar.

> I expect that discussion will continue for a while.

No no, please I don't want to be an obstacle here. I was just trying to fill the gap, sorry :).

bjnath commented:
> I was just trying to fill the gap, sorry :)

I'm very grateful for your help.  

Eager to see what you have in mind for the sidebar.

eric-wieser commented:
> If I'd started by trying to understand the theme and making changes bottom-up we'd have pieces all over the floor and still no PR.

My understanding is you can just drop your stylesheet changes in that repo. The mistake in #14648 seemed to be starting from scratch without any real buy-in for doing so.

rgommers commented:
> @rgommers: So by association, the https://github.com/scipy/scipy-sphinx-theme-v2 repo is defunct?

Yes, I don't think it's going to make it over the finish line:( As discussed in the last docs meeting, I think the scope of what @bjnath is doing here should be limited to making the style match the website a little better and fixing the most egregious issues. Eventually I think we would like a bigger makeover, but at this point I'd much rather adopt a good and well-maintained theme (e.g. one of the Pandas or Statsmodels ones) than keeping our own.

bjnath commented:
Closing because we now use a new theme (#17074)

Some of these changes have been migrated to that theme (#17382).
